compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -2216,14 +2222,31 @@ class Typer extends Namer
           case Some(xtree) =>
             traverse(xtree :: rest)
           case none =>
-            val defCtx = mdef match
+            def defCtx = ctx.withNotNullInfos(initialNotNullInfos)
+            val newCtx = if (ctx.owner.isTerm) {
               // Keep preceding not null facts in the current context only if `mdef`
               // cannot be executed out-of-sequence.
-              case _: ValDef if !mdef.mods.is(Lazy) && ctx.owner.isTerm =>
-                ctx // all preceding statements will have been executed in this case
```

Comment:
This is too complicated and also too inefficient. We should try to avoid unnecessary allocations here, and we should try to keep the code in typedStats reasonably short, since both factors affect runtime to a significant degree.
I propose to leave the code as is and just add a call to `adaptCreationContext(vdef)` in the first case clause. I.e.:
```scala
val defCtx = mdef match
  case vdef: ValDef if !mdef.mods.is(Lazy) && ctx.owner.isTerm =>
    // Keep preceding not null facts in the current context only if `mdef`
    // cannot be executed out-of-sequence.
    adaptCreationContext(vdef)
    ctx
  case _ =>
    ctx.withNotNullInfos(initialNotNullInfos)

Reply:
The reason we didn't check `ValDef` and `Lazy` in our code is because we want it works for general DefTree. As long as it is not forward referenced, it is ok to use the NotNullIInfos. We also don't need to worry about mutable variables used out of order here, since it checked by function `usedOutOfOrder` later.

For example:
```scala
val s: String|Null = ???
if (s != null) {
  def f = // it is safe to use the fact s!=null
}
```

```scala
val s: String|Null = ???
val s = f
if (s == null) throw Exception
def f = {
  // it is not safe to use the fact s!=null
  // when f is first used, we don't know the fact about s
}
```

Reply:
Ah, yes, good points. The we should use  `adaptCreationContext(mdef)` for all definitions.

Reply:
`adaptCreationContext(mdef: DefTree): Boolean` is added